libs/@blockprotocol/graph/src/stdlib/bound.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  NonNullTimeInterval,
+  TemporalBound,
+} from "../types/temporal-versioning.js";
+
+export const compareBounds = (
+  lhs: TemporalBound,
+  rhs: TemporalBound,
+  lhsType: keyof NonNullTimeInterval,
+  rhsType: keyof NonNullTimeInterval,
+): number => {
+  // If the bound values are not equal, then the bound with the lower value is less than the bound with the higher
+  // value.
+  if (lhs.kind !== "unbounded" && rhs.kind !== "unbounded") {
+    if (lhs.limit !== rhs.limit) {
+      return lhs.limit.localeCompare(rhs.limit);
+    }
+  }
+
+  if (lhs.kind === rhs.kind && lhsType === rhsType) {
+    return 0;
+  }
+
+  if (
+    (lhs.kind === "unbounded" && lhsType === "start") ||
+    (rhs.kind === "unbounded" && rhsType === "end") ||
+    (lhs.kind === "exclusive" &&
+      rhs.kind === "exclusive" &&
+      lhsType === "end" &&
+      rhsType === "start") ||
+    (lhs.kind === "exclusive" &&
+      rhs.kind === "inclusive" &&
+      lhsType === "end") ||
+    (lhs.kind === "inclusive" &&
+      rhs.kind === "exclusive" &&
+      rhsType === "start")
+  ) {
+    return -1;
+  }
+
+  if (
+    (lhs.kind === "unbounded" && lhsType === "end") ||
+    (rhs.kind === "unbounded" && rhsType === "start") ||
+    (lhs.kind === "exclusive" &&
+      rhs.kind === "exclusive" &&
+      lhsType === "start" &&
+      rhsType === "end") ||
+    (lhs.kind === "exclusive" &&
+      rhs.kind === "inclusive" &&
+      lhsType === "start") ||
+    (lhs.kind === "inclusive" && rhs.kind === "exclusive" && rhsType === "end")
+  ) {
+    return 1;
+  }
+  throw new Error(
+    `Implementation error, failed to compare bounds.\nLHS: ${JSON.stringify(
+      lhs,
+    )}\nRHS: ${JSON.stringify(rhs)}`,
+  );
+};
+
+/**
+ * Checks whether two given temporal bounds are adjacent to one another, where adjacency is defined as being next to one
+ * another on the timeline, without any points between, *and where they are not overlapping*.
+ *
+ * As such, two inclusive (or two exclusive) bounds with the same limit are *not* adjacent.
+ *
+ * @param lhs - The first bound of the comparison (order is unimportant)
+ * @param rhs - The second bound of the comparison
+ */
+export const boundIsAdjacentTo = (
```

Comment:
I'd also prefer if this is not exposed as this is an implementation detail of `intervalIsAdjacentToInterval` (which maybe should also be renamed to `IntervalsAreAdjacent`)

Reply:
> but I think with the documentation this function is fine as standalone.

I disagree. The function is only well defined, if the two bounds don't have the same kind, e.g. when comparing a start-bound with an end-bound and this is not stated in the docs. I don't know what it means if a start bound is adjacent to another start bound.

Reply:
Adjacence requires that the bounds are directed, and the directions of the bounds need to be different. I appreciate, that this function does not take the bound kind, but only because for the logic used in this scope it has some invariants, which don't need to be checked. If both bounds are of the same kind, it should always return false, but we know, that we will always call this function with two different bound kinds.

Reply:
Can you point me to any material online that defines what it means for two _bounds_ to be adjacent?

If you cannot (and I did have a search), then I'd pose that we are free to define this function however we want. And in the definition I have given its behaviour, which I have described in the documentation, there are no hidden invariants, it should _not_ return false if they're of the same kind.

That is why the _interval_ adjacency function does additional logic (such as only checking a `start` bound against and `end` bound)

Reply:
There is no definition for adjacency for bounds, only for intervals. To check for the adjacence of an interval you need to compare bounds, and this is exactly, what this function is doing. This is why this function **is** an implementation detail. It's only a helper function to avoid writing the same functionality twice.

Reply:
Right, which is why I'm suggesting we are free to define the operation as we wish, and _not_ encapsulate those additional invariants you describe

Reply:
I'm not going to block this PR in this comment, but I want to emphasize that I'm not sold on exposing this function as is

Reply:
Why it has to be inlined? Why can't you just not export it?

Reply:
Because they are in different modules, and you need to `export` to make it available in another module.

Reply:
Well, I consider this a language limitation, which could be solved by either moving the function next to the required function or at least adjusting the documentation to state, how to correctly use that function

Reply:
> could be solved by either moving the function next to the required function

Just bumping to highlight I edited my previous comment as I accidentally submitted it while typing.

> or at least adjusting the documentation to state, how to correctly use that function

As I said before, I am of the opinion that the documentation _does_ state what the function does and therefore how to correctly use it. The function's logic and return values make absolutely no presumptions about start/end bounds, etc. Those semantics are encapsulated by the _calling_ function.

It says:

> where adjacency is defined as being next to one another on the timeline, without any points between, *and where they are not overlapping*

That is what the function does. If in some strange world someone decided to check if two start bounds are adjacent (with this definition of adjacency), then that's their prerogative and the function would still be _correct_.

Reply:
Correct, the _interval_ overlaps, but that's not what we're discussing here. The bounds as limit points do not overlap.